config/configmapconverter.go

```diff
@@ -0,0 +1,52 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package config // import "go.opentelemetry.io/collector/config"
+
+import (
+	"context"
+)
+
+// MapConverter interface to convert a config.Map. This allows distributions
+// (or components) to build backwards compatible config converters.
+type MapConverter interface {
+	Convert(ctx context.Context, cfgMap *Map) error
+
+	// privateMapConverter is an unexported func to disallow direct implementation.
+	privateMapConverter()
```

Comment:
This seems overly-complex.  Why is implementation of this interface disallowed?  If that is necessary, why have `converter` as an intermediary rather than just adding the private method to `ConvertFunc`?

Reply:
Consistency with other places.

Reply:
> If that is necessary, why have converter as an intermediary rather than just adding the private method to ConvertFunc?

Because if I have `ConverterFunc` with the private method users can do `var cmc MapConverter = ConvertFunc(myFunc)` which defeats the purpose because if want to add new func on the interface I have to add it in the `ConvertFunc` as well which is not very clear.

Reply:
If you add new functions onto the interface you need to implement them on `converter` as well.  How is that any different?

In the Go SIG we made the decision to only use the private method pattern of preventing implementation where we do not want outside implementations.  In other words, interfaces that were for the SDK to implement and to be consumed by others.  This seems the opposite, something that is likely to be implemented by others and only consumed be the collector framework.

Reply:
...........

Here is simple, if we add a new function (indeed all new funcs have to have a "default" implementation), we will add them to the `converter` for sure, but all the current implementation that use `NewConverter` will continue to work.

Just give a try if you want to add a `CloseFunc`:
```golang
type MapConverter interface {
	Convert(ctx context.Context, cfgMap *Map) error

	Close(ctx context.Context) error

	privateMapConverter()
}

// ConvertFunc specifies the function invoked when the MapConverter.Convert is being called.
type ConvertFunc func(context.Context, *Map) error

// Convert implements the MapConverter.Convert.
func (f ConvertFunc) Convert(ctx context.Context, cfgMap *Map) error {
	if f == nil {
		return nil
	}
	return f(ctx, cfgMap)
}

// CloseFunc specifies the function invoked when the Retrieved.Close is being called.
type CloseFunc func(context.Context) error

// Close implements the Retrieved.Close.
func (f CloseFunc) Close(ctx context.Context) error {
	if f == nil {
		return nil
	}
	return f(ctx)
}

// MapConverterOption represents the possible options for NewMapConverter.
type MapConverterOption func(*retrieved)

// WithClose overrides the default `Close` function for a Retrieved.
// The default always returns nil.
func WithClose(closeFunc CloseFunc) MapConverterOption {
	return func(o *converter) {
		o.CloseFunc = closeFunc
	}
}

type converter struct {
	ConvertFunc
	CloseFunc
}

func (converter) privateMapConverter() {}

// NewMapConverter returns a MapConverter configured with the provided options.
func NewMapConverter(convertFunc ConvertFunc, options ...MapConverterOption) MapConverter {
	ret := &converter{
		ConvertFunc: convertFunc,
	}

	for _, op := range options {
		op(ret)
	}

	return ret
}
```

Reply:
> In the Go SIG we made the decision to only use the private method pattern of preventing implementation where we do not want outside implementations. In other words, interfaces that were for the SDK to implement and to be consumed by others. This seems the opposite, something that is likely to be implemented by others and only consumed be the collector framework.

This is a different SIG, and we went with different approaches, so we will be consistent with this SIG previous work.

Reply:
> < giant wall of code >
...

This isn't how interfaces in Go work.  This is excessively complicated for no apparent reason.  If you have a `Converter` interface:

```go
type Converter interface {
	Convert(ctx context.Context, cfgMap *Map) error
}
```

That can be implemented directly and used by the code that needs it.  No need for a private function or an unexported wrapper type or an initializer.  A `ConverterFunc` type could be added for convenience.

If you want to add a `Closer` interface, then do so:

```go
type Closer interface {
	Close(ctx context.Context) error
}
```

If a `Converter` needs to also be a `Closer`, it implements that interface.  The consumer can check if it is given a `Closer` and invoke `Close()` as necessary.  If the consumer, for some reason, really needs a `Converter` to also be guaranteed to be a `Closer`, it can define a `ConvertCloser` interface:

```go
type ConvertCloser interface {
	Converter
	Closer
}
```

That's it.  Nothing more.  No need to be more complicated than that.

Reply:
> This is a different SIG, and we went with different approaches, so we will be consistent with this SIG previous work.

As all of the changes you are making to the configuration system show, nothing is yet set in stone.  We can change prior decisions when we find that they were mistaken.  We did that in the Go SIG on this very topic.  It's okay.

Reply:
Please open an issue with the proposal, or a draft PR. Happy to review proposals, for the moment as I said I will keep things consistent, otherwise if I am the one who applies a new pattern I will have to change the pattern everywhere, and I will not do that, so happy to review things :).

FYI: I do know how the optional interfaces work in `go` :), but we decided not to go with that path since it complicates the usage especially when you have to check lots of optional implementations. This decision